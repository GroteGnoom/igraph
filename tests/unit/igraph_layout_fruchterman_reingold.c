/*
   IGraph library.
   Copyright (C) 2021  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>
#include "test_utilities.inc"

void make_box(int vertices, float half_size, igraph_vector_t *bounds) {
    for (int i = 0; i < 4; i++) {
        igraph_vector_init(&bounds[i], vertices);
    }
    igraph_vector_fill(&bounds[0], -half_size);
    igraph_vector_fill(&bounds[1], half_size);
    igraph_vector_fill(&bounds[2], -half_size);
    igraph_vector_fill(&bounds[3], half_size);
}

void destroy_bounds(igraph_vector_t *bounds) {
    for (int i = 0; i < 4; i++) {
        igraph_vector_destroy(&bounds[i]);
    }
}

int main() {
    igraph_t g;
    igraph_matrix_t result;
    igraph_vector_t bounds[4];
    igraph_vector_t weights;
    igraph_rng_seed(igraph_rng_default(), 42);

    printf("Empty graph:\n");
    igraph_small(&g, 0, 0, -1);
    igraph_matrix_init(&result, 0, 0);
    IGRAPH_ASSERT(igraph_layout_fruchterman_reingold(&g, &result, /*use_seed*/ 0,
                  /*niter*/ 0, /*start_temp*/ 0.0, IGRAPH_LAYOUT_NOGRID,
                  /*weight*/ NULL, /*minx*/ NULL, /*maxx*/ NULL, /*miny*/ NULL,
                  /*maxy*/ NULL) == IGRAPH_SUCCESS);
    print_matrix(&result);
    igraph_matrix_destroy(&result);
    igraph_destroy(&g);

    printf("Full graph of 10 vertices with loops and bounds:\n");
    igraph_full(&g, 10, 0, 1);
    igraph_matrix_init(&result, 0, 0);
    make_box(10, 1.0, bounds);
    IGRAPH_ASSERT(igraph_layout_fruchterman_reingold(&g, &result, /*use_seed*/ 0,
                  /*niter*/ 100, /*start_temp*/ 1.0, IGRAPH_LAYOUT_AUTOGRID,
                  /*weight*/ NULL, &bounds[0], &bounds[1], &bounds[2], &bounds[3]) == IGRAPH_SUCCESS);
    print_matrix(&result);
    igraph_matrix_destroy(&result);
    destroy_bounds(bounds);

    printf("Full graph of 10 vertices with loops and weights:\n");
    igraph_matrix_init(&result, 0, 0);
    igraph_vector_init(&weights, 55);
    igraph_vector_fill(&weights, 100);
    IGRAPH_ASSERT(igraph_layout_fruchterman_reingold(&g, &result, /*use_seed*/ 0,
                  /*niter*/ 100, /*start_temp*/ 1.0, IGRAPH_LAYOUT_AUTOGRID,
                  &weights, /*minx*/ NULL, /*maxx*/ NULL, /*miny*/ NULL,
                  /*maxy*/ NULL) == IGRAPH_SUCCESS);
    print_matrix(&result);
    igraph_matrix_destroy(&result);
    igraph_destroy(&g);
    igraph_vector_destroy(&weights);

    printf("Full graph of 5 vertices, seed and no iterations:\n");
    igraph_full(&g, 5, 0, 0);
    igraph_real_t seed[10] = {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0};
    matrix_init_real_row_major(&result, 5, 2, seed);
    IGRAPH_ASSERT(igraph_layout_fruchterman_reingold(&g, &result, /*use_seed*/ 1,
                  /*niter*/ 0, /*start_temp*/ 100, IGRAPH_LAYOUT_GRID,
                  /*weight*/ NULL, /*minx*/ NULL, /*maxx*/ NULL, /*miny*/ NULL,
                  /*maxy*/ NULL) == IGRAPH_SUCCESS);
    print_matrix(&result);
    igraph_matrix_destroy(&result);
    igraph_destroy(&g);

    VERIFY_FINALLY_STACK();
    return 0;
}
