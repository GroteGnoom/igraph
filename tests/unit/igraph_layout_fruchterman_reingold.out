Empty graph:
Full graph of 10 vertices with loops and bounds:
[ -0.560704 0.359932
  0.00323775 -0.648759
  0.725751 0.203469
   0.48278 -0.615813
  -0.443609 -0.569148
  0.338592 0.523258
  -0.615049 -0.12616
  -0.0526167 0.0517897
  -0.127928 0.656471
  0.513271 -0.175449 ]
Full graph of 10 vertices with loops and weights:
[ 0.306062 -0.725682
  0.195375 -0.623354
  0.188661 -0.932881
  0.204214 -0.836587
  0.382377 -0.788479
  0.310531 -0.635757
  0.309894 -0.874842
  0.0781742 -0.825945
  0.174828 -0.731794
  0.0833526 -0.693754 ]
Full graph of 5 vertices, seed and no iterations:
[      0.1      0.2
       0.3      0.4
       0.5      0.6
       0.7      0.8
       0.9        1 ]
